docs/site/Sequence.md

```diff
@@ -15,21 +15,72 @@ The contract of a `Sequence` is simple: it must produce a response to a request.
+  /**
+   * Optional invoker for registered middleware in a chain.
+   * To be injected via SequenceActions.INVOKE_MIDDLEWARE.
+   */
+  @inject(SequenceActions.INVOKE_MIDDLEWARE, {optional: true})
+  protected invokeMiddleware: InvokeMiddleware = () => false;
+
+  /**
+   * Constructor: Injects findRoute, invokeMethod & logError
+   * methods as promises.
+   *
+   * @param findRoute - Finds the appropriate controller method,
+   *  spec and args for invocation (injected via SequenceActions.FIND_ROUTE).
+   * @param parseParams - The parameter parsing function (injected
+   * via SequenceActions.PARSE_PARAMS).
+   * @param invoke - Invokes the method specified by the route
+   * (injected via SequenceActions.INVOKE_METHOD).
+   * @param send - The action to merge the invoke result with the response
+   * (injected via SequenceActions.SEND)
+   * @param reject - The action to take if the invoke returns a rejected
+   * promise result (injected via SequenceActions.REJECT).
+   */
+  constructor(
+    @inject(SequenceActions.FIND_ROUTE) protected findRoute: FindRoute,
+    @inject(SequenceActions.PARSE_PARAMS) protected parseParams: ParseParams,
+    @inject(SequenceActions.INVOKE_METHOD) protected invoke: InvokeMethod,
+    @inject(SequenceActions.SEND) public send: Send,
+    @inject(SequenceActions.REJECT) public reject: Reject,
+  ) {}
+
+  /**
+   * Runs the default sequence. Given a handler context (request and response),
+   * running the sequence will produce a response or an error.
+   *
+   * Default sequence executes these steps
+   *  - Executes middleware for CORS, OpenAPI spec endpoints
+   *  - Finds the appropriate controller method, swagger spec
+   *    and args for invocation
+   *  - Parses HTTP request to get API argument list
+   *  - Invokes the API which is defined in the Application Controller
+   *  - Writes the result from API into the HTTP response
+   *  - Error is caught and logged using 'logError' if any of the above steps
+   *    in the sequence fails with an error.
+   *
+   * @param context - The request context: HTTP request and response objects,
+   * per-request IoC container and more.
+   */
+  async handle(context: RequestContext): Promise<void> {
     try {
-      const route = this.findRoute(context.request);
-      const params = await this.parseParams(context.request, route);
-      const result = await this.invoke(route, params);
-      await this.send(context.response, result);
+      const {request, response} = context;
+      // Invoke registered Express middleware
+      const finished = await this.invokeMiddleware(context);
+      if (finished) {
+        // The response been produced by the middleware chain
+        return;
+      }
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      debug('%s result -', route.describe(), result);
+      this.send(response, result);
```

Comment:
What happens when `this.send` returns a promise that will be eventually rejected? IMO, we should either keep `await` or return the send result directly.

```suggestion
      return this.send(response, result);
```

Reply:
The signature of `Send` returns `void`:

```ts
export type Send = (response: Response, result: OperationRetval) => void;
```

Reply:
Interesting, I was remembering that both `Send` and `Reject` were returning a promise.

This change is fine then, just make sure to update other places for consistency please.